### PR TITLE
Add more examples

### PR DESCRIPTION
**Pull request by 4censord**
Adding some more examples.

- [x] centos8-netinstall
- [x] centos8-local
- [x] commented cento8 example. 
- [ ] ~~multistage build~~
- [ ] ~~xva-build~~

### CODE REVIEW

**Comment by dsko1234** [comment]
@ddelnano I cant get packer to agnollege that there exists an xva builder, have you tried xva before?
without getting the xva builder to work I cannot do a multistage build either

Any recomendationes?

**Comment by ddelnano** [comment]
@dsko1234 appreciate you opening up this. As mentioned in the README, the xva builder is not tested. So it's likely that it is broken. Could you create an issue for that?

**Comment by ddelnano** [comment]
@dsko1234 can you please rebase your changes now that #13 is merged? Also what do you think of dropping the xva build for now? I don't know how much work it will be to fix it and I think we shouldn't block this change as a result.

**Comment by dsko1234** [comment]
* Rebased against master
* Removed the centos json example
* added not on running the ubuntu example

**Comment by dsko1234** [comment]
I agree on droping the xva and mutlistage examples for now.

**Comment by dsko1234** [comment]
now that packer init is working with hashicorp/packer/pull/10760 
i would like to include packer init blocks into all examples.
also i might try upgrading the ubuntu examples aswell

**Comment by ddelnano** [comment]
> i would like to include packer init blocks into all examples.

That's fine with me but we will need to wait to merge this until Packer 1.7.1 is released (which will include the linked fix).

**Comment by ddelnano** [comment]
@4censord now that packer 1.7.1 is released could you rebase this again? This PR is also needed to close #14 since json templates are not supported with `packer init`.

Screen Shot 2021-04-13 at 12 53 56 PM

**Comment by 4censord** [comment]
Resolved all conflicts, and rebased against master.

File: docs/builders/iso/Commented-Example.md
Comment (ddelnano) [inline docs/builders/iso/Commented-Example.md:88]: I appreciate the detailed comments but it means that we will need to keep the `ks-centos8-examples.cfg` insync with this markdown file. Since kickstarter files support comments can we mention in this markdown file that the examples are well commented and move all of these into the kickstarter file?

File: docs/builders/iso/Commented-Example.md
Comment (ddelnano) [inline docs/builders/iso/Commented-Example.md:29]: Same comment as the kickstarter file. I am worried that this will get out of sync with the hcl file. Since this is no longer JSON we can provide the comments in the HCL file and mention that the example is well commented.

File: examples/README.md
Comment (ddelnano) [inline examples/README.md:24]: Placeholder comment to remind me to run these commands to make sure they work from scratch.

File: examples/centos/centos8-local.pkr.hcl
Comment (ddelnano) [inline examples/centos/centos8-local.pkr.hcl:4]: Please update all the examples to use a "real" release (v0.3.2). The `-dev` versions were for testing the new Github release process and shouldn't be used by an end user.

File: examples/ubuntu/ubuntu-2004.pkr.hcl
Comment (ddelnano) [inline examples/ubuntu/ubuntu-2004.pkr.hcl:71]: I know you are just copying this from the existing example, but it seems like this was accidentally duplicated

I'd like to remove this duplication if it's not necessary to get the example running.

File: examples/ubuntu/ubuntu-2004.pkr.hcl
Comment (ddelnano) [inline examples/ubuntu/ubuntu-2004.pkr.hcl:78]: Can we remove the `shutdown_command` from all the examples? Judging from #19, this is likely erroring out and relying on the XAPI vm shutdown api anyway. The docs were updated in that issue to indicate that this config key should be for exceptional cases only and that the default should work for most usecases.

**Review by ddelnano — CHANGES_REQUESTED** [review]
Really appreciate your work on this. I added some comments and have a few other questions.

What is the difference between the centos8-examples, centos8-netinstaller and centos8-local examples? It seems to me that the netinstaller and local ones install via cdrom vs a remote ISO (as seen below)
```
$ diff examples/http/centos8/ks-centos8-netinstall.cfg  examples/http/centos8/ks-centos8-local.cfg
5a6,7
> cdrom
>
10d11
< url --url="http://mirror.centos.org/centos/8.3.2011/BaseOS/x86_64/os/"
```

What I'm not understanding is the purpose of the centos8-example. It seems like the netinstaller example except that it includes the configuration from the original centos8 example implementation. Do we need all 3 or can the netinstaller and local ones suffice?

Once the comments are resolved, I'll want to run through the testing of all examples on the final commit before merging since we don't have an automated way to tests these.

File: examples/centos/centos8-local.pkr.hcl
Comment (4censord) [reply examples/centos/centos8-local.pkr.hcl:4]: Done

File: examples/ubuntu/ubuntu-2004.pkr.hcl
Comment (4censord) [reply examples/ubuntu/ubuntu-2004.pkr.hcl:78]: I didn't know about #19, will do

File: docs/builders/iso/Commented-Example.md
Comment (4censord) [reply docs/builders/iso/Commented-Example.md:88]: Done

**Comment by 4censord** [comment]
> What is the difference between the centos8-examples, centos8-netinstaller and centos8-local examples? It seems to me that the netinstaller and local ones install via cdrom vs a remote ISO 

That is correct, but it also use a different iso: centos8-netinstall(800Mb) vs centos8-dvd1(9Gb)
I added a cdrom install example after seeing a comment in the original kickstart file

```
# TODO: Figure out why cdrom does not work
# cdrom
```

> What I'm not understanding is the purpose of the centos8-example. It seems like the netinstaller example except that it includes the configuration from the original centos8 example implementation. Do we need all 3 or can the netinstaller and local ones suffice?

The netinstall and local are optimized for usability, ie export some values an run packer while the example is optimized for understanding/simplicity, so no variables and no pulling values from the environment.
As such is easier to understand and explain for someone not experienced with packer

If we were to get rid of one, i would recommend the local one.

File: .github/workflows/release.yml
Comment (ddelnano) [inline .github/workflows/release.yml:28]: Accidental change? I believe 1.16 was required during the packer-sdk upgrade.

File: docs/builders/iso/xenserver-iso.html.markdown
Comment (ddelnano) [inline docs/builders/iso/xenserver-iso.html.markdown:157]: Not sure where this came from but the `shutdown_command` is correctly documented on master so please remove.

File: docs/builders/iso/xenserver-iso.html.markdown
Comment (4censord) [inline docs/builders/iso/xenserver-iso.html.markdown:161]: Is shutdown_timeout still a thing?
I don't see it in master anymore, but it might still be usefull if one needs a shutdown command

Comment (ddelnano) [reply docs/builders/iso/xenserver-iso.html.markdown:161]: Upstream (github.com/xenserver/packer-builder-xenserver) had that config documented but I haven't seen any references aside from the documentation (https://github.com/xenserver/packer-builder-xenserver/search?q=shutdown_timeout). So I likely removed it because I believe the value is hardcoded [here](https://github.com/ddelnano/packer-plugin-xenserver/blob/a472fd2b02699a72f080ca08aaa483d6f27101e1/builder/xenserver/common/step_shutdown.go#L48).

File: examples/ubuntu/ubuntu-2004.pkr.hcl
Comment (ddelnano) [inline examples/ubuntu/ubuntu-2004.pkr.hcl:77]: This timeout was lowered from the previous `60000s` value and the `ssh_handshake_attempts` (which was set to 10000) was removed. This is causing the ubuntu example to fail with the following error

```
ddelnano@ddelnano-desktop:~/go/src/github.com/xenserver/packer-builder-xenserver$ packer1.7.1 build examples/ubuntu/ubuntu-2004.pkr.hcl
xenserver-iso.ubuntu-2004: output will be in this color.

==> xenserver-iso.ubuntu-2004: XAPI client session established
==> xenserver-iso.ubuntu-2004: Retrieving ISO
==> xenserver-iso.ubuntu-2004: Trying http://releases.ubuntu.com/20.04/ubuntu-20.04.2-live-server-amd64.iso
==> xenserver-iso.ubuntu-2004: Trying http://releases.ubuntu.com/20.04/ubuntu-20.04.2-live-server-amd64.iso?checksum=d1f2bf834bbe9bb43faf16f9be992a6f3935e65be0edece1dee2aa6eb1767423
==> xenserver-iso.ubuntu-2004: http://releases.ubuntu.com/20.04/ubuntu-20.04.2-live-server-amd64.iso?checksum=d1f2bf834bbe9bb43faf16f9be992a6f3935e65be0edece1dee2aa6eb1767423 => /home/ddelnano/go/src/github.com/xenserver/packer-builder-xenserver/packer_cache/dc8718d65b80640808a10b87f6d44fc2d16420af
==> xenserver-iso.ubuntu-2004: Starting HTTP server on port 8000
==> xenserver-iso.ubuntu-2004: Step: Upload VDI 'ubuntu-20.04.2-live-server-amd64.iso'
==> xenserver-iso.ubuntu-2004: Step: Found SRs '[OpaqueRef:beeed204-90ff-4f11-bfe2-1dfbb46d5eb5 OpaqueRef:ac11b102-c83c-4b0a-b076-2b5050474ac0 OpaqueRef:a6aca570-c039-4d21-8f2c-16c38456bdfd OpaqueRef:9963d6b4-d507-4425-90e6-2d857a25c4d0 OpaqueRef:90bdec93-5ba1-4187-bedd-5f18d3318410 OpaqueRef:10e7e21a-0569-4e6c-931e-d295a76a3be0]'
==> xenserver-iso.ubuntu-2004: PUT 'https://<sensitive>/import_raw_vdi?session_id=OpaqueRef%3A3a09798d-ccea-4da0-a069-3090f0dd42e2&task_id=OpaqueRef%3Ad5bae3ea-cf5d-4bf4-8dcd-f0089e966877&vdi=OpaqueRef%3Ada7653a0-d1d0-48fc-a777-11866f19fa56'
==> xenserver-iso.ubuntu-2004: Step: Create Instance
==> xenserver-iso.ubuntu-2004: Using the following SR for the VM: OpaqueRef:10e7e21a-0569-4e6c-931e-d295a76a3be0
==> xenserver-iso.ubuntu-2004: Created instance 'a16a1de7-eaee-c035-a349-add74a2018bd'
==> xenserver-iso.ubuntu-2004: Step: Start VM Paused
==> xenserver-iso.ubuntu-2004: Step: Set SSH address to VM host IP
==> xenserver-iso.ubuntu-2004: Set host SSH address to '192.168.88.89'.
==> xenserver-iso.ubuntu-2004: Unpausing VM a16a1de7-eaee-c035-a349-add74a2018bd
==> xenserver-iso.ubuntu-2004: Waiting 10s for boot...
==> xenserver-iso.ubuntu-2004: Connecting to the VM console VNC over xapi via 192.168.88.89
==> xenserver-iso.ubuntu-2004: Making HTTP request to initiate VNC connection: CONNECT /console?uuid=1e2a424b-b302-c212-a489-b177c49580f5 HTTP/1.0
==> xenserver-iso.ubuntu-2004: Host: 192.168.88.89
==> xenserver-iso.ubuntu-2004: Cookie: session_id=OpaqueRef:3a09798d-ccea-4da0-a069-3090f0dd42e2
==> xenserver-iso.ubuntu-2004:
==> xenserver-iso.ubuntu-2004:
==> xenserver-iso.ubuntu-2004: Received response: HTTP/1.1 200 OK
==> xenserver-iso.ubuntu-2004: Connection: keep-alive
==> xenserver-iso.ubuntu-2004: Cache-Control: no-cache, no-store
==> xenserver-iso.ubuntu-2004:
==> xenserver-iso.ubuntu-2004:
    xenserver-iso.ubuntu-2004: Found local IP: 192.168.88.254
==> xenserver-iso.ubuntu-2004: Typing boot commands over VNC...
==> xenserver-iso.ubuntu-2004: Finished typing.
==> xenserver-iso.ubuntu-2004: Step: Wait for VM's IP to become known to us.
    xenserver-iso.ubuntu-2004: Got IP '192.168.88.26' from HTTP request
==> xenserver-iso.ubuntu-2004: Got IP address '192.168.88.26'
==> xenserver-iso.ubuntu-2004: Creating a local port forward over SSH on local port 5900
==> xenserver-iso.ubuntu-2004: Port forward setup. 5900 ---> 192.168.88.26:22 on 192.168.88.89
==> xenserver-iso.ubuntu-2004: Using ssh communicator to connect: 192.168.88.26
==> xenserver-iso.ubuntu-2004: Waiting for SSH to become available...
==> xenserver-iso.ubuntu-2004: Error waiting for SSH: Packer experienced an authentication error when trying to connect via SSH. This can happen if your username/password are wrong. You may want to double-check your credentials as part of your debugging process. original error: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none password], no supported methods remain
==> xenserver-iso.ubuntu-2004: Deleting output directory...
Build 'xenserver-iso.ubuntu-2004' errored after 5 minutes 24 seconds: Packer experienced an authentication error when trying to connect via SSH. This can happen if your username/password are wrong. You may want to double-check your credentials as part of your debugging process. original error: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none password], no supported methods remain
```

I have not tested if adding the `ssh_handshake_attempts` is enough to get it working, but I was able to verify using the previous values work.

While it's crappy to have those values set very high to get the example to work, I think we should keep the values as is (unless you want to investigate setting them more properly now).

**Review by ddelnano — CHANGES_REQUESTED** [review]
> The netinstall and local are optimized for usability, ie export some values an run packer, while the example is optimized for understanding/simplicity, so no variables and no pulling values from the environment.

> As such is easier to understand and explain for someone not experienced with packer

I'd prefer to make all of the examples parameterized. While it may appear simpler, I think it's easier for a new packer user to be told exactly which environment variables to export and the correct command line arguments to run it with. I feel like it's more difficult for a user to read through the packer template to determine what is relevant to update.

> If we were to get rid of one, i would recommend the local one.

I like that the local one uses a different install method than the netinstall and centos8-example templates and since I prefer to have all the examples use variables it seems to me that the centos8-example should be removed, however, I'm fine with either option.

One con with the local template is that it takes a significant time to run. Does it have to use a 8GB ISO to use the `cdrom` kickstart command? I don't understand enough about how those ISOs are different and what options are available to.

I added a hopefully final round of feedback that I discovered as I tried to run the examples fully and thanks again for all your hard work on this PR!

File: examples/centos/centos8-netinstall.pkr.hcl
Comment (4censord) [reply examples/centos/centos8-netinstall.pkr.hcl:26]: done

File: examples/ubuntu/ubuntu-2004.pkr.hcl
Comment (4censord) [reply examples/ubuntu/ubuntu-2004.pkr.hcl]: Reverted to old values in 320e8de

**Comment by 4censord** [comment]
> I'd prefer to make all of the examples parameterized. While it may appear simpler, I think it's easier for a new packer user to be told exactly which environment variables to export and the correct command line arguments to run it with. I feel like it's more difficult for a user to read through the packer template to determine what is relevant to update.

I see your point, but personaly when i started writing my own packer files this might have helped me.

> I like that the local one uses a different install method than the netinstall and centos8-example templates and since I prefer to have all the examples use variables it seems to me that the centos8-example should be removed, however, I'm fine with either option.

I have no real preference either, so I have removed the centos-example.


> One con with the local template is that it takes a significant time to run. 
True, it may be a bit faster if one is using a local mirror, but still 8Gb is 8Gb.

> Does it have to use a 8GB ISO to use the cdrom kickstart command?

Yes.

> I don't understand enough about how those ISOs are different and what options are available to.

The DVD iso includes a full, up to date mirror of the centos 8 repositorys. The DVD iso is the one you would use for installs while having no/slow/expensive Internet.
The netboot ISO includes the minimum set of packages required to boot and install via a external Mirror

File: examples/README.md
Comment (ddelnano) [reply examples/README.md:24]: They all work.

**Comment by ddelnano** [comment]
Appreciate you creating the github issues for the known problems and your latest changes lgtm. Thanks again for all the effort on this change!

**Pull request merged**
